pkg/controllers/statusaggregator/plugins/job.go

```diff
@@ -0,0 +1,163 @@
+	"k8s.io/klog/v2"
+
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/common"
+)
+
+type JobPlugin struct{}
+
+func NewJobPlugin() *JobPlugin {
+	return &JobPlugin{}
+}
+
+func (receiver *JobPlugin) AggregateStatuses(
+	ctx context.Context,
+	sourceObject, fedObject *unstructured.Unstructured,
+	clusterObjs map[string]interface{},
+	clusterObjsUpToDate bool,
+) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("status-aggregator-plugin", "jobs")
+
+	needUpdate := false
+	aggregatedStatus := &batchv1.JobStatus{}
+	var completionTime *metav1.Time
+	var completionCount int
+	var jobCompleted, jobFailed []string
+
+	for clusterName, clusterObj := range clusterObjs {
+		logger := klog.FromContext(ctx).WithValues("cluster-name", clusterName)
+
+		utd := clusterObj.(*unstructured.Unstructured)
+		// For status of job
+		var found bool
+		status, found, err := unstructured.NestedMap(utd.Object, common.StatusField)
+		if err != nil || !found {
+			logger.Error(err, "Failed to get status of cluster object")
+			return nil, false, err
+		}
+
+		if status == nil {
+			continue
+		}
+
+		jobStatus := &batchv1.JobStatus{}
+		if err = runtime.DefaultUnstructuredConverter.FromUnstructured(status, jobStatus); err != nil {
+			logger.Error(err, "Failed to convert the status of cluster object")
+			return nil, false, err
+		}
+
+		if aggregatedStatus.StartTime.IsZero() || jobStatus.StartTime.Before(aggregatedStatus.StartTime) {
+			aggregatedStatus.StartTime = jobStatus.StartTime
+		}
+		if !jobStatus.CompletionTime.IsZero() {
+			completionCount++
+			jobCompleted = append(jobCompleted, clusterName)
+			if completionTime.IsZero() || completionTime.Before(jobStatus.CompletionTime) {
+				completionTime = jobStatus.CompletionTime
+			}
+		} else if IsJobFinishedWithFailed(jobStatus) {
+			completionCount++
+			jobFailed = append(jobFailed, clusterName)
+		}
+
+		aggregatedStatus.Active += jobStatus.Active
+		aggregatedStatus.Succeeded += jobStatus.Succeeded
+		aggregatedStatus.Failed += jobStatus.Failed
+	}
+	if completionCount > 0 && completionCount == len(clusterObjs) {
+		now := time.Now()
+		var conditionType batchv1.JobConditionType
+		var reason, message string
+		sort.Strings(jobCompleted)
+		sort.Strings(jobFailed)
+
+		switch {
+		case len(jobCompleted) > 0 && len(jobFailed) > 0:
+			conditionType = batchv1.JobFailed
+			reason = "Mixed"
+			message = fmt.Sprintf("Job completed in clusters [%s] and failed in member clusters [%s]",
+				strings.Join(jobCompleted, ","), strings.Join(jobFailed, ","))
+		case len(jobCompleted) > 0:
+			conditionType = batchv1.JobComplete
+			reason = "Completed"
+			message = fmt.Sprintf("Job completed in clusters [%s]", strings.Join(jobCompleted, ","))
+		default:
+			conditionType = batchv1.JobFailed
+			reason = "Failed"
+			message = fmt.Sprintf("Job failed in clusters [%s]", strings.Join(jobFailed, ","))
+		}
+
+		aggregatedStatus.CompletionTime = completionTime
```

Comment:
When there are failed jobs, should we set `CompletionTime`? https://github.com/kubernetes/kubernetes/blob/a17149e1a189050796ced469dbd78d380f2ed5ef/pkg/controller/job/job_controller.go#L568 The Job controller seems to only set `CompletionTime` when the Job succeeded.

Reply:
We should follow the semantics of the Job controller and only set `CompletionTime` when the job succeeds